group_theory/perm.lean

```diff
@@ -286,9 +385,204 @@ funext $ λ f,
+  ⟨-i, by rwa [gpow_neg, inv_gpow, inv_inv]⟩⟩
+
+lemma swap_mul_eq_mul_swap (f : perm α) (x y : α) : swap x y * f = f * swap (f⁻¹ x) (f⁻¹ y) :=
+equiv.ext _ _ $ λ z, begin
+  simp [mul_apply, swap_apply_def],
+  split_ifs;
+  simp [*, eq_inv_iff_eq] at * <|> cc
+end
+
+lemma mul_swap_eq_swap_mul (f : perm α) (x y : α) : f * swap x y = swap (f x) (f y) * f :=
+by rw [swap_mul_eq_mul_swap, inv_apply_self, inv_apply_self]
+
+lemma is_cycle_swap_mul_aux : ∀ (n : ℤ) {b x : α} {f : perm α} (hf : is_cycle f)
+  (hb : (swap x (f x) * f) b ≠ b)
+  (h : (f ^ n) (f x) = b), ∃ i : ℤ, ((swap x (f x) * f) ^ i) (f x) = b
+| 0         := λ b x f hf hb h, ⟨0, h⟩
+| (n+1 : ℕ) := λ b x f hf hb h,
+  if hfbx : f x = b then ⟨0, hfbx⟩ else
+  have f b ≠ b ∧ b ≠ x := support_swap_mul hb,
+  have hb : (swap x (f x) * f) (f⁻¹ b) ≠ f⁻¹ b,
+    by rw [mul_apply, apply_inv_self, swap_apply_of_ne_of_ne this.2 (ne.symm hfbx),
+        ne.def, ← injective.eq_iff f.bijective.1, apply_inv_self];
+      exact this.1,
+  have wf : n < int.nat_abs (1 + ↑n),
+    by rw [add_comm, ← int.coe_nat_one, ← int.coe_nat_add, int.nat_abs_of_nat];
+    exact nat.lt_succ_self n,
+  let ⟨i, hi⟩ := is_cycle_swap_mul_aux n hf hb
+    (f.bijective.1 $ by
+      rw [apply_inv_self];
+      rwa [gpow_coe_nat, pow_succ, mul_apply] at h) in
+  ⟨i + 1, by rw [add_comm, gpow_add, mul_apply, hi, gpow_one, mul_apply, apply_inv_self,
+      swap_apply_of_ne_of_ne this.2 (ne.symm hfbx)]⟩
+| -[1+ n] := λ b x f hf hb h,
+  if hfbx : f⁻¹ x = b then ⟨-1, by rwa [gpow_neg, gpow_one, mul_inv_rev, mul_apply, swap_inv, swap_apply_right]⟩
+  else if hfbx' : f x = b then ⟨0, hfbx'⟩
+  else
+  have f b ≠ b ∧ b ≠ x := support_swap_mul hb,
+  have hb : (swap x (f⁻¹ x) * f⁻¹) (f⁻¹ b) ≠ f⁻¹ b,
+    by rw [mul_apply, swap_apply_def];
+      split_ifs;
+      simp [inv_eq_iff_eq, eq_inv_iff_eq] at *; cc,
+  have wf : n < n.succ, from nat.lt_succ_self n,
+  let ⟨i, hi⟩ := is_cycle_swap_mul_aux n (is_cycle_inv hf) hb
+    (show (f⁻¹ ^ ↑n) (f⁻¹ x) = f⁻¹ b, by
+      rw [← h, ← mul_apply, ← mul_apply, ← mul_apply, gpow_neg_succ, ← inv_pow, pow_succ', mul_assoc,
+        mul_assoc, inv_mul_self, mul_one, gpow_coe_nat, ← pow_succ', ← pow_succ]) in
+  have h : (swap x (f⁻¹ x) * f⁻¹) (f x) = f⁻¹ x, by rw [mul_apply, inv_apply_self, swap_apply_left],
+  ⟨-i, by rw [← add_sub_cancel i 1, neg_sub, sub_eq_add_neg, gpow_add, gpow_one, gpow_neg, ← inv_gpow,
+      mul_inv_rev, swap_inv, mul_swap_eq_swap_mul, inv_apply_self, swap_comm _ x, gpow_add, gpow_one,
+      mul_apply, mul_apply (_ ^ i), h, hi, mul_apply, apply_inv_self, swap_apply_of_ne_of_ne this.2 (ne.symm hfbx')]⟩
+using_well_founded {rel_tac := λ _ _, `[exact ⟨_, measure_wf int.nat_abs⟩]}
+
+lemma eq_swap_of_is_cycle_of_apply_apply_eq_self {f : perm α} (hf : is_cycle f) {x : α}
+  (hfx : f x ≠ x) (hffx : f (f x) = x) : f = swap x (f x) :=
+equiv.ext _ _ $ λ y,
+have ∀ n : ℕ, (f ^ n) x = x ∨ (f ^ n) x = f x :=
```

Comment:
it would be nice to extract this and the next proof as separate lemmas for `f ^ n` with `f o f = id`.

Reply:
Not sure exactly what you mean bu this but I made some changes

Reply:
I thought it would be good to have the lemmas:
```lean
lemma pow_apply_eq_of_self_inverse_nat {f : perm α} (hffx : f (f x) = x) : ∀ n : ℕ, (f ^ n) x = x ∨ (f ^ n) x = f x := ...
lemma pow_apply_eq_of_self_inverse_int {f : perm α} (hffx : f (f x) = x) : ∀ i : ℤ, (f ^ i) x = x ∨ (f ^ i) x = f x := ...
```
(names are up for discussion...)

Reply:
I proved your first suggestion.
The generalization to groups is actually a weaker statement since `∃ x, f (f x) = x` does not imply `f * f = 1`.